### PR TITLE
What's new in .NET 7

### PR DESCRIPTION
**Pull request by gewarren**
Adds new article "What's new in .NET 7" - [preview link](https://review.learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-7?branch=pr-en-us-32113).

Note that this will merge to a release branch, not main.

Contributes to #29176

### CODE REVIEW

**Review by jkoritzinsky — APPROVED** [review]
Interop-related changes LGTM

File: docs/core/whats-new/dotnet-7.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+Much work has gone into Windows Forms for .NET 7. Improvements have been made in the following areas:
+
+- Accessibility
+- High DPI and scaling
+- Databinding
+
+For more information, see [What's new in Windows Forms in .NET 7](https://devblogs.microsoft.com/dotnet/winforms-enhancements-in-dotnet-7).
+
+<!--
+### Orleans
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:166]: I think you can uncomment this out, as long as we plan to release these at the same time. My Orleans article will be ready by then.

File: docs/core/whats-new/dotnet-7.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+To find all the <https://learn.microsoft.com/dotnet> articles that have been updated for .NET 7, see [What's new in docs for .NET 7](dotnet-7-docs.md).
+-->
+
+## Performance
+
+Performance is a key focus of .NET 7, and all of its features are designed with performance in mind. In addition, .NET 7 includes the following enhancements aimed purely at performance:
+
+- On-stack replacement (OSR) is a complement to tiered compilation. It allows the runtime to change the code executed by a currently running method in the middle of its execution (that is, while it's "on stack"). Long-running methods can switch to more optimized versions mid-execution.
+- Profile-guided optimization (PGO) now works with OSR and is easier to enable (by adding `<TieredPGO>true</TieredPGO>` to your project file). PGO can also instrument and optimize additional things, such as delegates.
+- Improved code generation for Arm64.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:25]:
```suggestion
- Improved code generation for ARM64.
```

I think think this is supposed to be in all caps.

File: docs/core/whats-new/dotnet-7.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+- [Native AOT](../deploying/native-aot/index.md) produces a standalone executable in the target platform's file format with no external dependencies. It's entirely native, with no IL or JIT, and provides fast startup time and a small, self-contained deployment. In .NET 7, Native AOT focuses on console apps and requires apps to be trimmed.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:26]:
```suggestion
- [Native AOT](../deploying/native-aot/index.md) produces a standalone executable in the target platform's file format with no external dependencies. It's entirely native, with no [IL or JIT](dotnet/core/introduction#compilation), and provides fast startup time and a small, self-contained deployment. In .NET 7, Native AOT focuses on console apps and requires apps to be trimmed.
```

Link to IL and JIT information.

File: docs/core/whats-new/dotnet-7.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+- Performance improvements to the mono runtime, which powers Blazor WebAssembly, Android, and iOS apps.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:27]:
```suggestion
- Performance improvements to the Mono Runtime, which powers Blazor WebAssembly, Android, and iOS apps.
```

I believe Mono Runtime is a proper name, https://www.mono-project.com/docs/advanced/runtime/

File: docs/core/whats-new/dotnet-7.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+For a detailed look at everything that's gone into making .NET 7 so fast, see the epic [Performance improvements in .NET 7](https://devblogs.microsoft.com/dotnet/performance_improvements_in_net_7/) blog post.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:29]:
```suggestion
For a detailed look at the majority of features that are focused solely on performance in .NET 7, see the [Performance improvements in .NET 7](https://devblogs.microsoft.com/dotnet/performance_improvements_in_net_7/) blog post.
```

The blog post itself states that it's not inclusive of everything, but rather a select 500 of the nearly 1,000 performance-specific PRs (7,000) PRs in total for .NET 7.

File: docs/core/whats-new/dotnet-7.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+- Support for **required members**, which are properties that must be present in the JSON payload for deserialization to succeed. For more information, see [Required properties](../../standard/serialization/system-text-json/required-properties.md).
+
+For information about these and other updates, see the [What's new in System.Text.Json in .NET 7](https://devblogs.microsoft.com/dotnet/system-text-json-in-dotnet-7/) blog post.
+
+## Generic math
+
+.NET 7 and C# 11 include innovations that allow you to perform mathematical operations generically&mdash;that is, without having to know the exact type you're working with. For example, if you wanted to write a method that adds two numbers, previously you had to add an overload of the method for each type. Now you can write a single, generic method, where the type parameter is constrained to be a number-like type. For more information, see [Generic math](../../standard/generics/math.md) and the [Generic math](https://devblogs.microsoft.com/dotnet/dotnet-7-generic-math/) blog post.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:43]:
```suggestion
.NET 7 and C# 11 include innovations that allow you to perform mathematical operations generically&mdash;that is, without having to know the exact type you're working with. For example, if you wanted to write a method that adds two numbers, previously you had to add an overload of the method for each type. Now you can write a single, generic method, where the type parameter is constrained to be a number-like type. For more information, see [Generic math](../../standard/generics/math.md) article and the [Generic math](https://devblogs.microsoft.com/dotnet/dotnet-7-generic-math/) blog post.
```

File: docs/core/whats-new/dotnet-7.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+  - <xref:System.Text.RegularExpressions.Regex.EnumerateMatches%2A?displayProperty=nameWithType>
+  - <xref:System.Text.RegularExpressions.Regex.Count%2A?displayProperty=nameWithType>
+  - <xref:System.Text.RegularExpressions.Regex.IsMatch(System.ReadOnlySpan{System.Char})?displayProperty=nameWithType> (and a few other overloads)
+
+For more information about these and other improvements, see the [Regular expression improvements in .NET 7](https://devblogs.microsoft.com/dotnet/regular-expression-improvements-in-dotnet-7/) blog post.
+
+## .NET libraries
+
+Many improvements have been made to .NET library APIs. Some are mentioned in other, dedicated sections of this article. Some others are summarized in the following table.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:65]:
```suggestion
Many improvements have been made to .NET library APIs. Some are mentioned in other, dedicated sections of this article. While others are summarized in the following table.
```

File: docs/core/whats-new/dotnet-7.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+- Available template names
+- Template options
+- Allowable option values
+
+In addition, for better conformity, the `install`, `uninstall`, `search`, `list`, and `update` subcommands no longer have the `--` prefix.
+
+#### Authoring
+
+Template *constraints*, a new concept for .NET 7, let you define the context in which your templates are allowed. Constraints help the template engine determine which templates it should show in commands like `dotnet new list`. You can constrain your template to an operating system, a template engine host (for example, the .NET CLI or New Project dialog in Visual Studio), and an installed workload. You define constraints in your template's configuration file.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:107]:
```suggestion
Template *constraints*, a new concept for .NET 7, lets you define the context in which your templates are allowed. Constraints help the template engine determine which templates it should show in commands like `dotnet new list`. You can constrain your template to an operating system, a template engine host (for example, the .NET CLI or New Project dialog in Visual Studio), and an installed workload. You define constraints in your template's configuration file.
```

File: docs/core/whats-new/dotnet-7.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+For more information, see [Central package management](/nuget/consume-packages/Central-Package-Management).
+
+## P/Invoke source generation
+
+.NET 7 introduces a source generator for platform invokes (P/Invokes) in C#. The source generator looks for <xref:System.Runtime.InteropServices.LibraryImportAttribute> on `static`, `partial` methods to trigger compile-time source generation of marshalling code. By generating the marshalling code at compile time, no IL stub needs to be generated at run time, as it does when using <xref:System.Runtime.InteropServices.DllImportAttribute>. The source generator improves application performance and also allows the app to be ahead-of-time (AOT) compiled. For more information, see [Source generation for platform invokes](../../standard/native-interop/pinvoke-source-generation.md) and [Use custom marshallers in source-generated P/Invokes](../../standard/native-interop/tutorial-custom-marshaller.md).
+
+## Related releases
+
+This section contains information about related products that have releases to coincide with the .NET 7 release.
```
Comment (IEvangelist) [inline docs/core/whats-new/dotnet-7.md:129]:
```suggestion
This section contains information about related products with releases that coincide with the .NET 7 release.
```

**Review by IEvangelist — COMMENTED** [review]
This is awesome, well done @gewarren! I have a few nits.

**Review by BillWagner — APPROVED** [review]
This all LGTM @gewarren 

One request for an addition: I'd recommend adding a link on the .NET Hub page, possibly one of the top cards.

File: docs/core/whats-new/dotnet-7.md (hunk 2 above)
Comment (gewarren) [reply docs/core/whats-new/dotnet-7.md:25]: It's contentious, but I'm in the Arm64 camp with Rich Lander and Tanner Gooding (and https://www.arm.com/).

File: docs/core/whats-new/dotnet-7.md (hunk 5 above)
Comment (gewarren) [inline docs/core/whats-new/dotnet-7.md:29]:
```suggestion
For a detailed look at many of the performance-focused features that make .NET 7 so fast, see the [Performance improvements in .NET 7](https://devblogs.microsoft.com/dotnet/performance_improvements_in_net_7/) blog post.
```

File: docs/core/whats-new/dotnet-7.md (hunk 7 above)
Comment (gewarren) [reply docs/core/whats-new/dotnet-7.md:65]: That makes it a sentence fragment so I think I'll keep it as-is.

File: docs/core/whats-new/dotnet-7.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+.NET 7 makes improvements to *observability*. Observability helps you understand the state of your app as it scales and as the technical complexity increases. .NET's observability implementation is primarily built around [OpenTelemetry](https://opentelemetry.io/). Improvements include:
+
+- The new <xref:System.Diagnostics.Activity.CurrentChanged?displayProperty=nameWithType> event, which you can use to detect when the span context of a managed thread changes.
+- New, performant enumerator methods for <xref:System.Diagnostics.Activity> properties: <xref:System.Diagnostics.Activity.EnumerateTagObjects>, <xref:System.Diagnostics.Activity.EnumerateLinks>, and <xref:System.Diagnostics.Activity.EnumerateEvents>.
+
+For more information, see the [.NET 7 Preview 4](https://devblogs.microsoft.com/dotnet/announcing-dotnet-7-preview-4/#observability) blog post.
+
+## .NET SDK
+
+The .NET 7 SDK improves the CLI template experience. It also enables publishing to containers and central package management with NuGet.
```
Comment (gewarren) [inline docs/core/whats-new/dotnet-7.md:89]:
```suggestion
The .NET 7 [SDK](../sdk.md) improves the CLI template experience. It also enables publishing to containers, and central package management with NuGet.
```

File: docs/core/whats-new/dotnet-7.md (hunk 8 above)
Comment (gewarren) [reply docs/core/whats-new/dotnet-7.md:107]: I'm reading it as "template contstraints...let you define".

File: docs/core/whats-new/dotnet-7.md (hunk 9 above)
Comment (gewarren) [inline docs/core/whats-new/dotnet-7.md:129]:
```suggestion
This section contains information about related products that have releases that coincide with the .NET 7 release.
```

Comment (gewarren) [reply docs/core/whats-new/dotnet-7.md:129]: I kept part of your suggestion, but I think "that have" is easier to translate than "with". With "with", the sentence could be taken to mean "This section contains info about related products. [This section] has releases to coincide..."

File: docs/core/whats-new/dotnet-7.md (hunk 1 above)
Comment (gewarren) [reply docs/core/whats-new/dotnet-7.md:166]: Do you want to target the `release-dotnet-7` branch for that PR?

**Comment by gewarren** [comment]
> One request for an addition: I'd recommend adding a link on the .NET Hub page, possibly one of the top cards.

@BillWagner It's already there (in this PR, that is):

image

**Pull request merged**